atlasdb-timelock-server/src/main/java/com/palantir/atlasdb/timelock/paxos/PaxosServerImplementation.java

```diff
@@ -0,0 +1,233 @@
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.atlasdb.timelock.paxos;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.stream.Collectors;
+
+import javax.net.ssl.SSLSocketFactory;
+
+import com.google.common.base.Optional;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+import com.google.common.net.HostAndPort;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import com.palantir.atlasdb.factory.Leaders;
+import com.palantir.atlasdb.http.AtlasDbHttpClients;
+import com.palantir.atlasdb.http.NotCurrentLeaderExceptionMapper;
+import com.palantir.atlasdb.timelock.ServerImplementation;
+import com.palantir.atlasdb.timelock.TimeLockServices;
+import com.palantir.atlasdb.timelock.config.PaxosConfiguration;
+import com.palantir.atlasdb.timelock.config.TimeLockServerConfiguration;
+import com.palantir.leader.LeaderElectionService;
+import com.palantir.leader.PaxosLeaderElectionService;
+import com.palantir.leader.PingableLeader;
+import com.palantir.leader.proxy.AwaitingLeadershipProxy;
+import com.palantir.lock.LockService;
+import com.palantir.lock.impl.LockServiceImpl;
+import com.palantir.paxos.PaxosAcceptor;
+import com.palantir.paxos.PaxosLearner;
+import com.palantir.paxos.PaxosProposer;
+import com.palantir.paxos.PaxosProposerImpl;
+import com.palantir.remoting.ssl.SslSocketFactories;
+import com.palantir.timestamp.PersistentTimestampService;
+import com.palantir.timestamp.TimestampService;
+
+import io.dropwizard.setup.Environment;
+
+public class PaxosServerImplementation implements ServerImplementation {
+    public static final String LEADER_NAMESPACE = "__leader";
+
+    private final PaxosConfiguration paxosConfiguration;
+    private final Environment environment;
+
+    private Set<String> remoteServers;
+    private Optional<SSLSocketFactory> optionalSecurity = Optional.absent();
+
+    private LeaderElectionService leaderElectionService;
+    private PaxosResource paxosResource;
+
+    public PaxosServerImplementation(PaxosConfiguration configuration, Environment environment) {
+        this.paxosConfiguration = configuration;
+        this.environment = environment;
+    }
+
+    @Override
+    public void onStartup(TimeLockServerConfiguration configuration) {
+        registerPaxosResource();
+
+        if (paxosConfiguration.sslConfiguration().isPresent()) {
```

Comment:
Good call. Amazingly this simplifies the code a lot (because of Guava's `transform()`) as well!